### ISSUE TITLE
DTLS Support - important for IKEA Trådfri home automation

### ISSUE THREAD

**Issue by edisno** [chrysn/aiocoap#67]
IKEA just released their "Trådfri" zigbee based light bulbs in various countries, and they supply a zigbee gateway which talks COAP+DTLS.  Some hackers who prefer python now use exec+libcoap which supports dtls out of the box.  Would be nice if dtls-enabled aiocoap could be the go-to library for this purpose :-)

**Comment by chrysn** [comment]
this is good news (awesome news would be the devices doing 6lowpan and coap themselves; maybe they need a little firmware help ;-) ).

there's still the open issue in pydtls that keeps me from using it (https://github.com/rbit/pydtls/issues/6 -- no 1.2 support that's required for coap), but maybe i can manage something with flextls.

have you successfully used the ikea devices with another library? libcoap (as of the latest version i see) doesn't support dtls; which version are you using? (it'd greatly help if i had another implementation to test against.)

**Comment by balloob** [comment]
Here is a tutorial to get IKEA devices working with libcoap https://community.home-assistant.io/t/ikea-tradfri-gateway-zigbee-very-basic-working-implementation/14788/19

**Comment by balloob** [comment]
It seems like pyDTLS is Python 2 only so that lib will have to be ported first.

**Comment by chrysn** [comment]
ad pyDTLS, yes, i noticed and opened https://github.com/rbit/pydtls/issues/11 to foster that (in light of comment number 2, *please* express "upvoting" of a bug only with the reaction ("thumbs up") feature, not with +1 comments unless the respective project's contribution policy explicitly says otherwise).

meanwhile, i'm looking into [flextls](https://github.com/DinoTools/python-flextls) -- it looks more basic and partially abandoned, but might still be suitable for initial experimental support (and if abandoned for good will hopefully replaced by pydtls once the abovementioned is done)

ad libcoap: thanks for the link, that should make getting started much easier.

**Comment by chrysn** [comment]
This is harder than expected; state of things:

* I've toyed around with flextls, but apart from a sever lack of documentation, this seems not to have an actual DTLS implementation, only parts thereof (parsing, partial state machine, but no full handshake implementation).
* I've made some progress with a Python 3 port of pydtls, only to discover that https://bugs.python.org/issue19084 is another blocker (pydtls relies on Python's ssl module, and that doesn't support PSK right now) -- the cipher that works in libcoap is TLS_PSK_WITH_AES_128_CCM_8 (0xc0a8), and that's not in Python's list.

The way flextls is used looks like it could have been applied to the message handling with the current socket infrastructure (one socket used for incoming and outgoing traffic with PKTINFO); pydtls looks more like it sits atop a socket and is either client or server. For coaps, such a setup might actually be viable for a correct implementation, as there's no multicast involved.

Options right now are shaving the complete yak (ie. adding PSK support to Python and following through with the Python3 port of pydtls), finding another library that does it out of the box, or using CFFI to crudely tap into libssl like I do for OSCOAP.

**Comment by meinarsson** [comment]
There seems to me some DTLS support for aiocoap based on tinydtls in this repo:  https://git.fslab.de/jkonra2m/aiocoapMC

**Comment by chrysn** [comment]
Thanks, that looks promising. I don't know yet whether the cython-tinydtls-wrapper can be a long-term solution, but right now I'm a bit stuck with pyDTLS (I quick-and-dirtied Python 3 and [OpenSSL 1.1 support](https://github.com/rbit/pydtls/issues/13) in, but still didn't get AES-CCM to run), and this might provide a quick solution that'd live in a pluggable transport.

**Comment by chrysn** [comment]
There is now an experimental (read: might rebase) tinydtls branch. It has various constraints from being DTLS-client-only over having the credentials hardcoded in the source to not detecting epoch changes, but it's a start.

**Comment by balloob** [comment]
It works with IKEA!

For people following this issue, to try at home:

 1. Use the tinydtls branch of this repo
 2. Follow [install instructions](https://github.com/chrysn/aiocoap/blob/tinydtls/aiocoap/transports/tinydtls.py#L12)
 3. Hardcode your IKEA key [here](https://github.com/chrysn/aiocoap/blob/tinydtls/aiocoap/transports/tinydtls.py#L84)
 4. Run `python3 setup.py develop` in aiocoap dir
 5. Run following script:

```python
import asyncio
import sys

# This is path from step 2
sys.path.insert(0, '/usr/src/app/tinydtls/cython')

from aiocoap import *
import dtls


@asyncio.coroutine
def main():
    # Replace IP with your IKEA IP
    msg = Message(code=GET, uri="coaps://192.168.1.19:5684/15001")
    protocol = yield from Context.create_client_context()
    res = yield from protocol.request(msg).response
    print("RECEIVED STATUS", res.code)
    print("RECEIVED PAYLOAD", res.payload.decode('utf-8'))


asyncio.get_event_loop().run_until_complete(main())
```

I get the following output:

```
decrypt_verify(): found 24 bytes cleartext
decrypt_verify(): found 46 bytes cleartext
RECEIVED STATUS 2.05 Content
RECEIVED PAYLOAD [65536,65537,65538,65539,65540]
Exception ignored in: 'dtls._write'
Traceback (most recent call last):
  File "/usr/src/app/aiocoap/aiocoap/transports/tinydtls.py", line 114, in _write
AttributeError: 'DTLSClientConnection' object has no attribute '_transport'
```

**Comment by chrysn** [comment]
On Tue, May 02, 2017 at 10:25:42PM -0700, Paulus Schoutsen wrote:
> It works with IKEA!

Awesome! Thanks for testing.

(Too bad though from the "I should have bought one just to see whether
it works *cough*" point of view ;-) )

> ```
> decrypt_verify(): found 24 bytes cleartext
> decrypt_verify(): found 46 bytes cleartext
> ```

I'm still working with the python-tinydtls author on the packaging and
library, so expect the building instructions to change; also, I hope to
get rid of those messages tinydtls prints.

> ```
> Exception ignored in: 'dtls._write'
> Traceback (most recent call last):
>   File "/usr/src/app/aiocoap/aiocoap/transports/tinydtls.py", line 114, in _write
> AttributeError: 'DTLSClientConnection' object has no attribute '_transport'
> ```

A workaround for that is pending upload.

**Comment by balloob** [comment]
We use the following configure flag for `libcoap` and that got us rid of those cleartext warnings (it was contributed to pytradfri so I'm not sure which part disables it)

```
./configure --disable-documentation --disable-shared --without-debug CFLAGS="-D COAP_DEBUG_FD=stderr"
```

**Comment by Terris-Dianomic** [comment]
Stupid question perhaps.. Can I use the tinydtls branch to make a coaps/dtls server?

**Comment by chrysn** [comment]
> Stupid question perhaps.. Can I use the tinydtls branch to make a coaps/dtls server?

Right now the code in aiocoap only supports the client case; I'm
currently trying to make that less rough first (no segfaults when server
goes away or 64k messages were sent) before I'd tackle the server side.

Feel free to play around, in aiocoap/transport/tinydtls.py to implement
create_server_transport_endpoint, but expect it to be a tough ride.

**Comment by chrysn** [comment]
There has been some progress on the DTLS branch today; still no server, but it should not crash any more when the server disappears, and reconnect when the server disconnects.

It still times out when the server is restarted quickly without sending a shutdown message, but that appears to me right now to be the (libcoap+tinydtls) server's fault as it doesn't send any alert in response to DTLS application data sent from a nonexistant session (that needs further investigation, but might better be done when there's an aiocoap DTLS server too).

**Comment by rfuentess** [comment]
Good day, 

I have been working with a DTLS Server side, which for now seems to be functional (yet, the amount of data and the number of sensors tested is very low). 

Sadly, I had problems trying to link callbacks to the `TransportEndpointTinyDTLS` and for now my testing are over the UDP endpoints. This is the main reason I still don't do a PR, yet the testing branch is [here](https://github.com/rfuentess/aiocoap/tree/tinydtls_server_v1).

**Comment by chrysn** [comment]
Finally hooking up aiocoap with a real Trådfri and added a first patch around an odd behavior (75680de3). Overall still experimental, though.

**Comment by chrysn** [comment]
The tinydtls branch has now been merged into master. Keeping this open until DTLS is at least as usable as to allow proper configuration of keys, and maybe server support.

**Comment by balloob** [comment]
This is awesome! Thanks

**Comment by balloob** [comment]
Would it be possible to push a new version to PyPi ?

**Comment by chrysn** [comment]
The tinydtls branch has now been merged into master. Keeping this open until DTLS is at least as usable as to allow proper configuration of keys, and maybe server support.
Ad PyPI, wondering where the comment I remember writing about that vanished: I'll update PyPI when there is a 0.4 version of which I'm confident it can be considered somewhat stable. If there is a way to upload packages to PyPI that are flagged as pre-release and not automatically pulled by everyone who says they want aiocoap, then I can upload there, but otherwise I suggest you stay with installation from git until 0.4.

**Comment by balloob** [comment]
PyPi supports this by having the version be indicated as a pre-release, ie `0.4.0a0`.

More info:
https://pip.pypa.io/en/stable/reference/pip_install/#pre-release-versions
https://pip.pypa.io/en/stable/reference/pip_install/#install-pre

**Comment by lwis** [comment]
@chrysn any news on a pre release going out for 0.4?

**Comment by chrysn** [comment]
@lwis, @balloob: still no 0.4, but I've just published the current state as 0.4a1.

**Comment by chrysn** [comment]
Closing this to keep things focused; follow-up issues are #97 on the credentials store and #98 on server-side DTLS.
